File: rd-net/RdFramework/Impl/MessageBroker.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

namespace JetBrains.Rd.Impl
{
  public class MessageBroker
  {
    private readonly ILog myLogger = Log.GetLog("protocol.Mq");

    public bool BackwardsCompatibleWireFormat = false;

    private readonly object myLock = new();
    private readonly Dictionary<RdId, ValueLifetimed<IRdWireable>> mySubscriptions = new();
    private Queue<byte[]>? myUnprocessedMessages;
    
    public MessageBroker(bool withholdMessageDeliveryInitially = false)
    {
      myUnprocessedMessages = withholdMessageDeliveryInitially ? new() : null;
    }

    public void StartDeliveringMessages()
    {
      while (true)
      {
        Queue<byte[]>? queue;
        lock (myLock)
        {
          queue = myUnprocessedMessages;
          Assertion.Require(queue != null, "Already started delivering messages");

          if (queue.Count == 0)
          {
            myUnprocessedMessages = null;
            return;
          }

          myUnprocessedMessages = new Queue<byte[]>();
        }

        foreach (var message in queue)
          DispatchImpl(message);
      }
    }

    //on poller thread
    public void Dispatch(byte[] msg)
    {
      if (myUnprocessedMessages != null)
      {
        lock (myLock)
        {
          if (myUnprocessedMessages is { } queue)
          {
            queue.Enqueue(msg);
            return;
          }
        }
      }

      DispatchImpl(msg);
    }

    private unsafe void DispatchImpl(byte[] msg)
    {
      fixed (byte* p = msg)
      {
        var reader = UnsafeReader.CreateReader(p, msg.Length);
        var id = RdId.Read(reader);

        if (!TryGetById(id, out var lifetimed))
        {
          myLogger.Trace($"Handler is not found for {id}");
          return;
        }

        var reactive = lifetimed.Value.NotNull();
        var proto = reactive.TryGetProto();
        if (proto == null)
        {
          myLogger.Trace($"proto is null for {id}");
          return;
        }

        using (AllowBindCookie.Create())
        {
          using var _ = UsingDebugInfoCookie(reactive);
          var messageContext = BackwardsCompatibleWireFormat ? default : proto.Contexts.ReadContextsIntoCookie(reader);
          var dispatcher = new RdWireableDispatchHelper(lifetimed.Lifetime, myLogger, reactive, id, proto, messageContext);
          reactive.OnWireReceived(reader, dispatcher);
          
        }

        reader.Reset(null, 0);
      }
    }

    private class RdWireableDispatchHelper : IRdWireableDispatchHelper
    {
      private readonly ILog myLog;
      private readonly IRdWireable myWireable;
      private readonly IProtocol myProtocol;
      private readonly ProtocolContexts.MessageContext myMessageContext;
      
      public Lifetime Lifetime { get; }
      public RdId RdId { get; }

      internal RdWireableDispatchHelper(Lifetime lifetime, ILog log, IRdWireable wireable, RdId rdId, IProtocol protocol, ProtocolContexts.MessageContext messageContext)
      {
        Lifetime = lifetime;
        myLog = log;
        myWireable = wireable;
        RdId = rdId;
        myProtocol = protocol;
        myMessageContext = messageContext;
      }

      public void Dispatch(IScheduler? scheduler, Action action)
      {
        DoDispatch(Lifetime, scheduler ?? myProtocol.Scheduler, action);
      }
      
      private void DoDispatch(Lifetime lifetime, IScheduler scheduler, Action action)
      {
        if (lifetime.IsNotAlive)
        {
          myLog.Trace($"Lifetime: {lifetime} is not alive for {RdId}");
          return;
        }
        
        myLog.Trace($"Schedule continuation for {RdId}");
        
        
        scheduler.Queue(() =>
        {
          if (lifetime.IsNotAlive)
          {
            myLog.Trace($"Lifetime: {lifetime} is not alive for {RdId}");
            return;
          }
          
          using (UsingDebugInfoCookie(myWireable))
          using (myMessageContext.UpdateCookie())
          {
            action();
          }
        });
      }
    }

    private static FirstChanceExceptionInterceptor.ThreadLocalDebugInfo UsingDebugInfoCookie(IRdWireable reactive)
    {
      return new FirstChanceExceptionInterceptor.ThreadLocalDebugInfo(reactive);
    }

    public void Advise(Lifetime lifetime, IRdWireable reactive)
    {
      var rdId = reactive.RdId;
      if (rdId.IsNil)
      {
        if (lifetime.IsNotAlive)
          return;
        
        Assertion.Fail($"!id.IsNil: {reactive}");
      }
      
      mySubscriptions.BlockingAddUnique(lifetime, myLock, rdId, new(lifetime, reactive));
    }

    public bool TryGetById(RdId rdId, out ValueLifetimed<IRdWireable> subscription)
    {
      lock (myLock)
        return mySubscriptions.TryGetValue(rdId, out subscription) && subscription.Lifetime.IsAlive;
    }
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/MessageBroker.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.framework.base.AllowBindingCookie
import com.jetbrains.rd.framework.base.IRdWireable
import com.jetbrains.rd.framework.base.IRdWireableDispatchHelper
import com.jetbrains.rd.framework.impl.ProtocolContexts
import com.jetbrains.rd.util.Sync
import com.jetbrains.rd.util.blockingPutUnique
import com.jetbrains.rd.util.error
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.isAlive
import com.jetbrains.rd.util.lifetime.isNotAlive
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.string.IPrintable
import com.jetbrains.rd.util.string.PrettyPrinter
import com.jetbrains.rd.util.trace
import java.util.concurrent.LinkedBlockingQueue

class RdMessage (val id: RdId, val istream: AbstractBuffer, val tooBigMessage: String?)
class MessageBroker(queueMessages: Boolean = false) : IPrintable {

    companion object {
        val log = Protocol.sublogger("MQ")
    }

    private var unprocessedMessages: LinkedBlockingQueue<RdMessage>? = if (queueMessages) LinkedBlockingQueue() else null

    private val lock = Any()

    private val subscriptions = hashMapOf<RdId, Entry>()

    fun startDeliveringMessages() {
        while (true) {
            val queue: LinkedBlockingQueue<RdMessage>
            Sync.lock (lock) {
                queue = requireNotNull(unprocessedMessages) { "Already started delivering messages" }

                if (queue.size == 0)
                {
                    unprocessedMessages = null
                    return
                }

                unprocessedMessages = LinkedBlockingQueue()
            }

            for (rdMessage in queue) {
                dispatchImpl(rdMessage.id, rdMessage.istream, rdMessage.tooBigMessage)
            }
        }
    }

    fun tryGetById(rdId: RdId): IRdWireable? = tryGetEntryById(rdId)?.subscription

    private fun tryGetEntryById(rdId: RdId): Entry? {
        val value = Sync.lock(lock) { subscriptions[rdId] } ?: return null
        if (value.lifetime.isAlive)
            return value
        return null
    }

    //only on poller thread
    fun dispatch(id: RdId, buffer: AbstractBuffer, tooBigMessage: String?) {
        require(!id.isNull) { "id mustn't be null" }

        if (unprocessedMessages != null) {
            Sync.lock(lock) {
                val queue = unprocessedMessages ?: return@lock
                queue.add(RdMessage(id, buffer, tooBigMessage))
                return
            }
        }

        dispatchImpl(id, buffer, tooBigMessage)
    }

    private fun dispatchImpl(id: RdId, buffer: AbstractBuffer, tooBigMessage: String?) {
        val entry = tryGetEntryById(id)
        if (entry == null) {
            log.trace { "handler is not found for $id" }
            return
        }

        val protocol = entry.subscription.protocol
        if (protocol == null) {
            log.trace { "protocol is not found for $id" }
            return
        }

        try {
            AllowBindingCookie.allowBind {
                val messageContext = protocol.contexts.readContext(buffer)
                val helper = RdWireableDispatchHelper(entry.lifetime, id, protocol, messageContext)
                val subscription = entry.subscription
                if (tooBigMessage != null) {
                    log.error { tooBigMessage + " Location: " + subscription.location.toString() }
                }
                subscription.onWireReceived(buffer, helper)
            }
        } catch (e: Throwable) {
            log.error("Unexpected exception happened during processing a protocol event", e)
        }
    }

    private class RdWireableDispatchHelper(
        override val lifetime: Lifetime,
        override val rdId: RdId,
        private val protocol: IProtocol,
        private val messageContext: ProtocolContexts.MessageContext
    ) : IRdWireableDispatchHelper {

        override fun dispatch(scheduler: IScheduler?, action: () -> Unit) {
            doDispatch(lifetime, scheduler ?: protocol.scheduler, action)
        }

        private fun doDispatch(lifetime: Lifetime, scheduler: IScheduler, action: () -> Unit) {
            if (lifetime.isNotAlive) {
                log.trace { "Lifetime: $lifetime is not alive for $rdId" }
                return
            }

            scheduler.queue {
                if (lifetime.isNotAlive) {
                    log.trace { "Lifetime: $lifetime is not alive for $rdId" }
                    return@queue
                }

                messageContext.update {
                    action()
                }
            }
        }
    }

    fun adviseOn(lifetime: Lifetime, entity: IRdWireable) {
        if (entity.rdid.isNull) {
            if (lifetime.isNotAlive)
                return

            error("id is null for entity: $entity")
        }

        subscriptions.blockingPutUnique(lifetime, lock, entity.rdid, Entry(lifetime, entity))
    }

    override fun print(printer: PrettyPrinter) {
        Sync.lock(lock) {
            printer.println("MessageBroker Dump")
            printer.println("Subscribers:\n  ")
            printer.println(subscriptions.entries.joinToString("  "))
        }
    }

    private class Entry(val lifetime: Lifetime, val subscription: IRdWireable)
}
